File: src/modules/graphs/5_cycleDetection_Directed/DFS.ts
export class Solution {
    private dfsCheck(node: number, adj: number[][], visited: number[], pathVisited: number[]): boolean {
        visited[node] = 1;
        pathVisited[node] = 1;

        // traverse for adjacency nodes
        for (const neighbor of adj[node]) {
            // node is not visited
            if (!visited[neighbor]) {
                if (this.dfsCheck(neighbor, adj, visited, pathVisited) === true) return true;
            } else if (pathVisited[neighbor]) {// if the node has been previoulsy visited, but it has to be visited on the same path
                return true;
            }
        }


        pathVisited[node] = 0;
        return false;
    }
    public isCyclic(adj: number[][]) {
        let n = adj.length
        let visited = new Array(n).fill(0);
        let pathVisited = new Array(n).fill(0);

        for (let i = 0; i < n; i++) {
            if (!visited[i]) {
                if (this.dfsCheck(i, adj, visited, pathVisited) === true) return true;
            }
        }
        return false;
    }
}

File: src/modules/graphs/5_cycleDetection_Directed/DFS.go
package main

import "fmt"

type StackItem struct {
	node  int
	start bool
}

type Solution struct{}

func (s *Solution) isCyclicUtil(node int, adj [][]int, visited []int, pathVisited []int) bool {
	stack := []StackItem{{node, true}}

	for len(stack) > 0 {
		current := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if current.start {
			visited[current.node] = 1
			pathVisited[current.node] = 1

			// Add a marker to reset pathVisited after exploring the node
			stack = append(stack, StackItem{current.node, false})

			// Add all adjacent nodes to the stack for DFS
			for _, neighbor := range adj[current.node] {
				if visited[neighbor] == 0 {
					stack = append(stack, StackItem{neighbor, true})
				} else if pathVisited[neighbor] == 1 {
					return true
				}
			}
		} else {
			pathVisited[current.node] = 0
		}
	}

	return false
}

func (s *Solution) isCyclic(adj [][]int) bool {
	n := len(adj)
	visited := make([]int, n)
	pathVisited := make([]int, n)

	for i := 0; i < n; i++ {
		if visited[i] == 0 {
			if s.isCyclicUtil(i, adj, visited, pathVisited) {
				return true
			}
		}
	}
	return false
}

func main() {
	s := &Solution{}
	adj := [][]int{
		{1},
		{2},
		{3},
		{4},
		{0},
	}

	fmt.Println(s.isCyclic(adj)) // Output: true
}
